# Guided Project: Exploring Ebay Car Sales Data

### In this guided project, I will work work with a dataset of used cars from eBay Kleinanzeigen, a classified section of the German eBay website

##### The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data)

### The aim of this project is to clean the data and analyze the included used car listings

There were made a few modifications from the original dataset:

- There were sampled 50,000 data points from the full dataset, to ensure the code runs quickly in our hosted environment
- The dataset was dirtied a bit to more closely resemble what we would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

**The data dictionary provided with data is as follows:**

`dateCrawled` - When this ad was first crawled. All field-values are taken from this date.

`name` - Name of the car.

`seller` - Whether the seller is private or a dealer.

`offerType` - The type of listing

`price` - The price on the ad to sell the car.

`abtest` - Whether the listing is included in an A/B test.

`vehicleType` - The vehicle Type.

`yearOfRegistration` - The year in which the car was first registered.

`gearbox` - The transmission type.

`powerPS` - The power of the car in PS.

`model` - The car model name.

`kilometer` - How many kilometers the car has driven.

`monthOfRegistration` - The month in which the car was first registered.

`fuelType` - What type of fuel the car uses.

`brand` - The brand of the car.

`notRepairedDamage` - If the car has a damage which is not yet repaired.

`dateCreated` - The date on which the eBay listing was created.

`nrOfPictures` - The number of pictures in the ad.

`postalCode` - The postal code for the location of the vehicle.

`lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos= pd.read_csv("autos.csv", encoding= "Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#### What do we know so far

From the information about the `autos` dataframe, we can see that:
    - The dataset contains 20 columns, most of which are strings 
    - Column names use camelcase instead of Python's preferred snakecase. I cannot replace the spaces with underscores.
    - `price` column has comma and $ sign. 
        - The latter definitely needs to be removed and the column label renamed to indicate that the price is in dollars
    - `vehicleType, gearbox, model, fuelType` and  `notRepairedDamage` columns have missing values `NaN`. The null values are not more than ~20%
    - `odometer` column contains numeric data and string `km`


### Cleaning column names

I will convert the column names from camelcase to snakecase. I will also do some rewording on the column names based on the data dictionary to be more descriptive.

In [6]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
autos.columns= [
"date_crawled",
"name",
"seller",
"offer_type",
"price",
"ab_test",
"vechile_type",
"registration_year",
"gearbox",
"power_ps",
"model",
"odometer",
"registration_month",
'fuel_type',
"brand",
"unrepaired_damage",
"ad_created",
"nr_pictures",
"post_code",
"last_seen"]

In [8]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In the cell above, I have converted the column names from camelcase to snakecase, to respect the Python's preferred writing syntax.
Moreover, I have brought small changes to the column names to be more descriptive of the data.

### Initial exploring and cleaning

Now it is time to do some basic data exploration to determine what other cleaning tasks need to be done. Initially, I will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- A second important step is to look for examples of numeric data stored as text which can be cleaned and converted.

In [9]:
autos.describe(include= "all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vechile_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures     post_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In [10]:
print(autos["nr_pictures"].value_counts())
print('\n')
print(autos["ab_test"].value_counts())
print('\n')
print(autos["seller"].value_counts())
print('\n')
print(autos["offer_type"].value_counts())

0    50000
Name: nr_pictures, dtype: int64


test       25756
control    24244
Name: ab_test, dtype: int64


privat        49999
gewerblich        1
Name: seller, dtype: int64


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


Making use of the `autos.describe(include= 'all')` method, I inpsected the columns and concluded the following:

- `nr_pictures` column can be dropped for the analysis because the only value it has is `0.0`, deeming it irrelevant for our analysis

- `ab_test` column can be dropped as well

- `registratrion_year` would need to be inspected more because be min and max values are 1000 and 9999 which is fairly impossible for a car's registration years

- `power_ps` min value is 0 which is very unlikely for a car to have zero horsepower

- `registration_month` min value is also 0
- `seller` and `offer_type` columns have the same values
- `date_crawled` could be transformed into a datetime object

For the columns `price` and `odometer`, which are stored as text, will need to convert them to numerical data. I will:

- remove any non-numeric characters
- convert the column to a numeric type
- rename the `odometer` column to `odometer_km`

In [11]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [12]:
autos["price"]= autos["price"].str.replace('$', "").str.replace(',', "").astype(int)
autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [13]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [14]:
autos["odometer"]= autos["odometer"].str.replace('km', "").str.replace(',', "").astype(int)
autos["odometer"]

autos.rename({"odometer": "odometer_km"}, axis= 1, inplace= True)

In [15]:
autos= autos.drop(["nr_pictures", "ab_test"], axis= 1)

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int64
vechile_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
post_code             50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(6), object(12)
memory usage: 6.9+ MB


In [17]:
autos.describe(include= "all")

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       vechile_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created     post_code  \
count              40171                50000  50000.000000   
unique                 2                   76           NaN   
top                 nein  2016-04-03 00:00:00           NaN   
freq               35232                 1946           NaN   
mean                 NaN                  NaN  50813.627300   
std                  NaN                  NaN  25779.747957   
min                  NaN                  NaN   1067.000000   
25%                  NaN                  NaN  30451.000000   
50%                  NaN                  NaN  49577.000000   
75%                  NaN                  NaN  71540.000000   
max                  NaN                  NaN  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Exploring the Odometer and Price Columns

I have converted `price` and `odometer` to numeric types and renamed `odometer` to `odometer_km`

For the begining , these columns need further attention as there are some values that seem unrealistic while using the `DataFrame.describe()` method and they might be required to be removed in order to not jeopardize the analysis.

In [18]:
print(autos["price"].unique().shape)

(2357,)


In [19]:
print(autos["price"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [20]:
print(autos["price"].value_counts().sort_index(ascending= True))

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

By inspecting the series `autos["price"]`, I can see that there are 1421 values corresponding to 0 and other values that are up to the price of 500.

Some of the prices are unrealistic fo some used cars such as `99999999` or `12345678`, `10000000` even if they are high-end brands. For this, I made some simple boolean indexing to show that the prices are unrealistic

In [21]:
autos.loc[autos["price"]== 99999999, "brand"]

39705    mercedes_benz
Name: brand, dtype: object

In [22]:
autos.loc[autos["price"]== 500, "brand"]

148          volkswagen
283      sonstige_autos
320             peugeot
371              nissan
433                 bmw
541                opel
542       mercedes_benz
652                ford
675               honda
694                ford
756          volkswagen
778          volkswagen
819                 kia
842                 bmw
864          volkswagen
897                opel
898          volkswagen
924                opel
926                ford
956               honda
979                audi
1205         volkswagen
1223            renault
1416         volkswagen
1438            renault
1509             toyota
1539               ford
1627               fiat
1641           chrysler
1678            citroen
              ...      
48284            lancia
48305              opel
48312             smart
48405        volkswagen
48413              fiat
48421               bmw
48437        volkswagen
48559              audi
48646              opel
48663        volkswagen
48699           

I will remove the 0 values and also the values above 999990, thus I will be left with a more realistic set of data regarding the pricing for the cars.

In [23]:
autos= autos[autos["price"].between(500, 350000)]

In [24]:
print(autos["price"].describe())

count     45097.000000
mean       6320.659600
std        9261.841444
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      350000.000000
Name: price, dtype: float64


In [25]:
print(autos["price"].unique().shape)

(2208,)


I have been left only with 2208 or rows regarding the price, from the total of 2357. This can mean that there is plenty of data in this column to support our analysis when we will get there.

In [26]:
print(autos["odometer_km"].unique().shape)

(13,)


In [27]:
print(autos["odometer_km"].describe())

count     45097.000000
mean     125293.035013
std       39622.744927
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [28]:
print(autos["odometer_km"].value_counts().sort_index(ascending= True))

5000        616
10000       237
20000       727
30000       765
40000       808
50000       996
60000      1131
70000      1189
80000      1385
90000      1676
100000     2031
125000     4838
150000    28698
Name: odometer_km, dtype: int64


So far, the `odometer_km` seem to have realistic data and I will got go any further to clean it.

### Exploring the date columns


There are 5 columns that represent the date values taking from the crawler of from the website itself:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

In [29]:
# Thought that this can be used to transform the date columns in datetime objects but it is not necessary
# `autos[["date_crawled", "ad_created", "last_seen", "registration_year", "registration_month"]]= autos[["date_crawled", "ad_created", "last_seen", "registration_year", "registration_month"]].astype("datetime64[ns]")`


Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [30]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

It is important to notice that the first 10 characters in every column represent the year, month and day. I would like to sort the dates in order to get a date range for the sales of the cars.

I will extract the date values and use `Series.value_counts()` to generate a distribution which I will then sort by the index.

First, I will select the first 10 characters in each column

In [31]:
autos["date_crawled"].str[:10].shape

(45097,)

In [32]:
autos["date_crawled"].str[:10].value_counts(normalize= True, dropna= False).sort_index(ascending= True)

2016-03-05    0.025567
2016-03-06    0.014125
2016-03-07    0.036189
2016-03-08    0.033173
2016-03-09    0.032907
2016-03-10    0.032707
2016-03-11    0.033018
2016-03-12    0.037320
2016-03-13    0.015522
2016-03-14    0.036300
2016-03-15    0.034016
2016-03-16    0.029359
2016-03-17    0.031155
2016-03-18    0.012883
2016-03-19    0.034747
2016-03-20    0.038073
2016-03-21    0.037741
2016-03-22    0.033018
2016-03-23    0.032397
2016-03-24    0.028982
2016-03-25    0.031089
2016-03-26    0.032641
2016-03-27    0.031177
2016-03-28    0.034836
2016-03-29    0.033262
2016-03-30    0.033328
2016-03-31    0.031665
2016-04-01    0.033905
2016-04-02    0.035767
2016-04-03    0.038827
2016-04-04    0.036610
2016-04-05    0.013172
2016-04-06    0.003171
2016-04-07    0.001353
Name: date_crawled, dtype: float64

We are looking at 45097 entry points for these dates.

The data crawled was for approximately 30 days. It can be seen that in the beginning of April 2016 the data crawled has drastically went down compared with March 2016

In [33]:
autos["ad_created"].str[:10].shape

(45097,)

In [34]:
autos["ad_created"].str[:10].value_counts(normalize= True, dropna= False).sort_index(ascending= True)

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
2015-12-30    0.000022
2016-01-03    0.000022
2016-01-07    0.000022
2016-01-10    0.000044
2016-01-13    0.000022
2016-01-14    0.000022
2016-01-16    0.000022
2016-01-22    0.000022
2016-01-27    0.000067
2016-01-29    0.000022
2016-02-01    0.000022
2016-02-02    0.000044
2016-02-05    0.000044
2016-02-07    0.000022
2016-02-08    0.000022
2016-02-09    0.000022
2016-02-11    0.000022
2016-02-12    0.000044
2016-02-14    0.000044
2016-02-16    0.000022
2016-02-17    0.000022
2016-02-18    0.000044
2016-02-19    0.000067
2016-02-20    0.000044
2016-02-21    0.000044
                ...   
2016-03-09    0.032996
2016-03-10    0.032441
2016-03-11    0.033328
2016-03-12    0.037098
2016-03-13    0.016963
2016-03-14    0.034880
2016-03-15    0.033794
2016-03-16    0.029847
2016-03-17    0.030822
2016-03-18    0.013504
2016-03-19    0.033616
2016-03-20    0.038207
2016-03-21 

We are following 45097 listings created between June 2015 up to April 2016.

From June 2015 and up until 21st of February 2016, the frequency for creating a selling post was very low: between 0,0022% - 0,0067%. Some strange distribution

In [35]:
autos["last_seen"].str[:10].shape

(45097,)

In [36]:
autos["last_seen"].str[:10].value_counts(normalize= True, dropna= False).sort_index(ascending= True)

2016-03-05    0.001087
2016-03-06    0.004169
2016-03-07    0.005211
2016-03-08    0.007007
2016-03-09    0.009468
2016-03-10    0.010289
2016-03-11    0.012041
2016-03-12    0.023904
2016-03-13    0.008870
2016-03-14    0.012285
2016-03-15    0.015677
2016-03-16    0.016165
2016-03-17    0.027674
2016-03-18    0.007406
2016-03-19    0.015411
2016-03-20    0.020423
2016-03-21    0.020667
2016-03-22    0.021243
2016-03-23    0.018405
2016-03-24    0.019536
2016-03-25    0.018582
2016-03-26    0.016476
2016-03-27    0.015456
2016-03-28    0.020534
2016-03-29    0.021354
2016-03-30    0.024148
2016-03-31    0.023438
2016-04-01    0.022862
2016-04-02    0.024880
2016-04-03    0.024946
2016-04-04    0.024303
2016-04-05    0.126616
2016-04-06    0.225314
2016-04-07    0.134155
Name: last_seen, dtype: float64

It seems that more users have been last seen in April. It might be that right then they have sold the car they listed

In [37]:
autos["registration_year"].describe()

count    45097.000000
mean      2005.064173
std         89.652017
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

For the `registration_year` we find two very peculiar dates:
- the minum value, also the earliest registration date, is the year 1000 (before cars were invented)
- the maximum value, also the most recent registration date, is the year 9999 which is in the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate, since the data is from 2016. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

I will explore the number of listings that are outside 1900 - 2016 interval and see if I can remove the entire row or if I need to write a custom code this.

In [38]:
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      19
1961       6
1962       4
1963       8
1964      11
1965      17
1966      21
        ... 
2000    2715
2001    2520
2002    2419
2003    2657
2004    2683
2005    2864
2006    2666
2007    2270
2008    2206
2009    2078
2010    1586
2011    1617
2012    1307
2013     797
2014     652
2015     364
2016     950
2017    1296
2018     462
2019       1
2800       1
4100       1
4500       1
4800       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 93, dtype: int64

After exploring the series `autos['registration_year']`, I decided to select the interval 1950 - 2016. I will make use of the `df[df['col_name'].between(x, y)]` method and to update the autos dataframe

In [39]:
autos= autos[autos["registration_year"].between(1950, 2016)]

In [40]:
autos["registration_year"].value_counts(normalize= True).sort_index()

1950    0.000023
1951    0.000046
1952    0.000023
1953    0.000023
1954    0.000046
1955    0.000046
1956    0.000092
1957    0.000046
1958    0.000092
1959    0.000139
1960    0.000439
1961    0.000139
1962    0.000092
1963    0.000185
1964    0.000254
1965    0.000393
1966    0.000485
1967    0.000600
1968    0.000600
1969    0.000439
1970    0.000785
1971    0.000577
1972    0.000716
1973    0.000531
1974    0.000554
1975    0.000416
1976    0.000485
1977    0.000508
1978    0.000970
1979    0.000785
          ...   
1987    0.001570
1988    0.002979
1989    0.003579
1990    0.006327
1991    0.006766
1992    0.007043
1993    0.007851
1994    0.011546
1995    0.019859
1996    0.024754
1997    0.034914
1998    0.046783
1999    0.059391
2000    0.062693
2001    0.058191
2002    0.055858
2003    0.061354
2004    0.061954
2005    0.066134
2006    0.061562
2007    0.052418
2008    0.050940
2009    0.047984
2010    0.036623
2011    0.037339
2012    0.030181
2013    0.018404
2014    0.0150

Majority of the cars were registered starting with the 1990's, with a very big occurence during the interval 1998-2009

In [41]:
autos["brand"].describe()

count          43306
unique            40
top       volkswagen
freq            9179
Name: brand, dtype: object

There is an occurence where there is a brand called `sonstige_autos` but from my research this represents other type of brands that may not be relevant or the owner did not tag his car with a brand name.

Since this is not relevant to my analysis in finding data about specific brands, I will drop off the rows with these tags.

In [42]:
find_row= autos[autos["brand"] == "sonstige_autos"].index
autos= autos.drop(find_row)

In [43]:
autos["brand"].value_counts(normalize= True) * 100

volkswagen       21.399762
bmw              11.614949
mercedes_benz    10.323363
opel             10.129858
audi              9.141352
ford              6.574499
renault           4.431959
peugeot           2.965519
fiat              2.375679
seat              1.806822
skoda             1.732217
smart             1.527056
mazda             1.515399
nissan            1.515399
citroen           1.429138
toyota            1.354533
hyundai           1.025808
mini              0.946541
volvo             0.946541
honda             0.788007
mitsubishi        0.781013
kia               0.741380
alfa_romeo        0.669107
porsche           0.648124
chevrolet         0.608491
suzuki            0.596834
chrysler          0.363696
dacia             0.286760
jeep              0.247127
land_rover        0.228476
daihatsu          0.219150
subaru            0.200499
saab              0.167860
jaguar            0.163197
daewoo            0.135220
rover             0.125895
trabant           0.111906
l

In [44]:
auto_brands= autos["brand"].value_counts().head(10).index
auto_brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')

In [45]:
brand_mean_prices= {}

for b in auto_brands:
    selected_rows= autos[autos["brand"] == b]
    sorted_rows= selected_rows.sort_values("price", ascending= False)
    mean= sorted_rows["price"].mean().round()
    brand_mean_prices[b] = int(mean)

brand_mean_prices

{'audi': 9614,
 'bmw': 8582,
 'fiat': 3256,
 'ford': 4278,
 'mercedes_benz': 8764,
 'opel': 3395,
 'peugeot': 3361,
 'renault': 2820,
 'seat': 4811,
 'volkswagen': 5780}

I can see that Mini, Audi, Mercedes Benz and BMW are among the most expensive cars, while at the other end there is Renault, Fiat, Peugeot and Opel.

The most popular car brand, Volkswagen, is in between as price making it an affordable car for medium-income people.

In [46]:
brand_mean_mileage= {}

for b in auto_brands:
    selected_rows= autos[autos["brand"] == b]
    sorted_rows= selected_rows.sort_values("odometer_km", ascending= False)
    mean= sorted_rows["odometer_km"].mean().round()
    brand_mean_mileage[b]= int(mean)
    
brand_mean_mileage

{'audi': 128910,
 'bmw': 132891,
 'fiat': 114416,
 'ford': 123619,
 'mercedes_benz': 131064,
 'opel': 128036,
 'peugeot': 126073,
 'renault': 126415,
 'seat': 120058,
 'volkswagen': 128238}

In [47]:
bmp_series= pd.Series(brand_mean_prices)
bmp_series

audi             9614
bmw              8582
fiat             3256
ford             4278
mercedes_benz    8764
opel             3395
peugeot          3361
renault          2820
seat             4811
volkswagen       5780
dtype: int64

In [50]:
bmp_df= pd.DataFrame(bmp_series, columns= ["mean_price"])
bmp_df

mean_price
audi                 9614
bmw                  8582
fiat                 3256
ford                 4278
mercedes_benz        8764
opel                 3395
peugeot              3361
renault              2820
seat                 4811
volkswagen           5780

In [51]:
bmm_series= pd.Series(brand_mean_mileage)
bmp_df["mean_mileage"]= bmm_series

In [52]:
bmp_df

mean_price  mean_mileage
audi                 9614        128910
bmw                  8582        132891
fiat                 3256        114416
ford                 4278        123619
mercedes_benz        8764        131064
opel                 3395        128036
peugeot              3361        126073
renault              2820        126415
seat                 4811        120058
volkswagen           5780        128238